app/webdav/server/methods/addWebdavAccount.js

```diff
@@ -31,24 +31,73 @@ Meteor.methods({
 			}
 		);
 
-		try {
-			await client.stat('/');
-		} catch (error) {
-			return { success: false, message: 'could-not-access-webdav', error };
-		}
-
 		const accountData = {
 			user_id: userId,
 			server_url: formData.serverURL,
 			username: formData.username,
 			password: formData.pass,
 			name: formData.name,
 		};
+
 		try {
+			await client.stat('/');
 			WebdavAccounts.insert(accountData);
-			return { success: true, message: 'webdav-account-saved' };
 		} catch (error) {
-			return { success: false, message: error.code === 11000 ? 'duplicated-account' : 'unknown-write-error', error };
+			if (error.code === 11000) {
+				throw new Meteor.Error('duplicated-account', {
+					method: 'addWebdavAccount',
+				});
+			}
+			throw new Meteor.Error('could-not-access-webdav', {
+				method: 'addWebdavAccount',
+			});
+		}
+		return true;
+	},
+
+	async addWebdavAccountByToken(data) {
+		const userId = Meteor.userId();
+
+		if (!userId) {
+			throw new Meteor.Error('error-invalid-user', 'Invalid User', { method: 'addWebdavAccount' });
+		}
+
+		if (!settings.get('Webdav_Integration_Enabled')) {
+			throw new Meteor.Error('error-not-allowed', 'WebDAV Integration Not Allowed', { method: 'addWebdavAccount' });
+		}
+
+		check(data, Match.ObjectIncluding({
+			serverURL: String,
+		}));
+
+		const client = createClient(data.serverURL, { token: data.token });
+
+		const accountData = {
+			user_id: userId,
+			server_url: data.serverURL,
+			token: data.token,
+			name: data.name,
+		};
+
+		try {
+			await client.stat('/');
+			WebdavAccounts.upsert({
+				user_id: userId,
+				name: data.name,
+			}, {
+				$set: accountData,
+			});
+		} catch (error) {
+			if (error.code === 11000) {
```

Comment:
why don't you check before trying to insert/update instead of checking the error code?